### DISCUSSION TITLE
Definition of Fabric Native Components using Swift

### DISCUSSION THREAD

**Discussion by flokol120** [reactwg/react-native-new-architecture#146] [Q&A] [answered]
After implementing a turbo native module using swift (#143), I am now trying to implement a Fabric Native Component using Swift.

Unfortunately, [RNNewArchitectureLibraries](https://github.com/react-native-community/RNNewArchitectureLibraries) does not include an example of a Fabric Native Componente written in Swift.

That's why I wanted to ask if this is possible before digging deeper into this. I already tried to implement the `RCT<ComponentName>ViewProtocol` Protocol within my custom view Swift class, however by doing this I am not able to build the iOS app anymore. So I am guessing this is not the right way to do this.

Using the legacy native components I was able to code my custom view class using Swift and just instantiate a new instance in the `view` function of my `RCTViewManager`-class. But as Turbo Native Components need to implement the generated ViewProtocol this doesn't seem to be possible anymore.

Is there any other way to achieve this?

**Comment by cipolleschi** [comment, marked as answer]
Hi @flokol120! Unfortunately, currently it is not possible to implement a Fabric Component using Swift easily.
The Fabric interfaces relies a lot on C++ code which is not usable for Swift as they not interop well enough for our purposes.

There could be wayt to do that, by adding an ObjectiveC++ layer in between, but we never tested it and the process could be cumbersome and introduce subtle bugs, so we are not advising this, nor working on that right now.

We have ideas on how to simplify the setup to enable using Swift to create Fabric Components, but currently is not our highest priority. :(

**Comment by aliraza-noon** [comment]
hi folks, what is the latest on this can we use fabric with swift now using the new architecture ?

**Reply by cipolleschi** [reply]
Hi @aliraza-noon. Sadly no, it's not yet possible to create a Fabric component only using Swift. The reason is that the APIs of Fabric components are mainly in C++ and there is no good interoperability between the two languages. Specifically, you would have to create a Swift class that inherit from an Objective-C++ class with some methods in C++ and this kind of inheritance is not supported by the Swift/C++ interoperability yet. 

The best approach is to create a Componente in Objective-C that inherits from the C++ one, compose it with the Swift implementation and forward the various method calls from the Objective-C wrapper to the Swift component.

Something like:

* Compnoent Header File
```objc
// inheritance
@interface RCTYourComponent: RCTViewComponentView
@end
```

* Compnent Implementation file:
```objc
@implementation RCTYourComponent {
  YourSwiftComponent * _component;
}

// init the swift component and add it as subview/contentview of the fabric component

-(void)updateProps:()oldProps newProps()newProps
{
  auto oldTypedProps = //cast the old props to the right type
  auto newTypedProps = //cast the old props to the right type

  // forward the update props to the Swift implementation
  [_component updateProps:oldTypedProps newProps:newTypedProps];

  [super updateProps:oldTypedProps newProps:newTypedProps];
}

@end
```

**Comment by vladimirivanoviliev** [comment]
For future reference - such example of having Fabric component written in Swift with ObjectiveC++ interop layer is available in the following discussion:
- https://github.com/reactwg/react-native-new-architecture/discussions/8#discussioncomment-12790617
